pkg/autodiscover/autodiscover_operators.go

```diff
@@ -63,32 +64,61 @@ func isIstioServiceMeshInstalled(allNs []string) bool {
 	return true
 }
 
-func findOperatorsByLabel(olmClient clientOlm.Interface, labels []labelObject, namespaces []configuration.Namespace) []*olmv1Alpha.ClusterServiceVersion {
-	csvs := []*olmv1Alpha.ClusterServiceVersion{}
-	for _, ns := range namespaces {
-		for _, aLabelObject := range labels {
-			label := aLabelObject.LabelKey + "=" + aLabelObject.LabelValue
-			log.Debug("Searching CSVs in namespace %q with label %q", ns, label)
-			csvList, err := olmClient.OperatorsV1alpha1().ClusterServiceVersions(ns.Name).List(context.TODO(), metav1.ListOptions{
-				LabelSelector: label,
-			})
-			if err != nil {
-				log.Error("Error when listing csvs in namespace %q with label %q", ns, label)
-				continue
-			}
-
-			for i := range csvList.Items {
-				csvs = append(csvs, &csvList.Items[i])
-			}
+func findLabelSelector(labels []labelObject) map[string]string {
+	labelSelector := make(map[string]string)
+	if len(labels) > 0 {
+		for _, l := range labels {
+			labelSelector[l.LabelKey] = l.LabelValue
+		}
+	}
+	return labelSelector
+}
+
+func findalloperators(olmClient clientOlm.Interface, labelSelector map[string]string, namespace configuration.Namespace) (*olmv1Alpha.ClusterServiceVersionList, error) {
+	var csvList *olmv1Alpha.ClusterServiceVersionList
+	var err error
+	csvList = &olmv1Alpha.ClusterServiceVersionList{}
+	for key, value := range labelSelector {
+		log.Debug("Searching CSVs in namespace %q with label selector %q", namespace, ""+key+"="+value)
+		csv, errList := olmClient.OperatorsV1alpha1().ClusterServiceVersions(namespace.Name).List(context.TODO(), metav1.ListOptions{
+			LabelSelector: key + "=" + value,
+		})
+		if errList != nil {
+			log.Error("Error when listing pods in ns=%s, err: %v", namespace, errList)
+			err = multierror.Append(err, errList)
 		}
+		csvList.Items = append(csvList.Items, csv.Items...)
 	}
+	return csvList, err
+}
+func findOperatorsByLabel(olmClient clientOlm.Interface, labels []labelObject, namespaces []configuration.Namespace) (csvs []*olmv1Alpha.ClusterServiceVersion, err error) {
+	csvs = []*olmv1Alpha.ClusterServiceVersion{}
+	var csvList *olmv1Alpha.ClusterServiceVersionList
 
+	labelSelector := findLabelSelector(labels)
+	for _, ns := range namespaces {
+		var errList error
+		if len(labelSelector) > 0 {
+			csvList, errList = findalloperators(olmClient, labelSelector, ns)
+		} else {
+			log.Debug("Searching CSVs in namespace %s without label selector", ns)
+			csvList, errList = olmClient.OperatorsV1alpha1().ClusterServiceVersions(ns.Name).List(context.TODO(), metav1.ListOptions{})
+		}
+		if errList != nil {
+			log.Error("Error when listing csvs in namespace %q , err: %v", ns, errList)
+			err = multierror.Append(err, errList)
```

Comment:
why use multierror? simply return err

Reply:
If it fails, I don't stop the loop. instead, I save the error and proceed to the next namespace, checking for each one. Finally, I return all the errors I've collected. That's why I utilized multierror. @edcdavid

Reply:
After discussion with David , I understand that we don't need "multierror".